src/vset.h

```diff
@@ -0,0 +1,312 @@
+ *-----------------------------------------------------------------------------
+ *
+ * Entries are grouped into **time buckets** based on their expiry time.
+ * Each time bucket represents a window aligned to:
+ *
+ *     #define VOLATILESET_BUCKET_INTERVAL_MIN  (1 << 4)  // 16ms
+ *     #define VOLATILESET_BUCKET_INTERVAL_MAX  (1 << 13) // 8192ms
+ *
+ * A time bucket key is computed by rounding the expiry timestamp up to the
+ * nearest aligned window using `get_bucket_ts()`.
+ *
+ *-----------------------------------------------------------------------------
+ * Entry Addition and Bucket Promotion
+ *-----------------------------------------------------------------------------
+ *
+ * When a new entry is added:
+ *
+ * 1. If the current set is `NONE`, it becomes a `SINGLE` bucket.
+ * 2. If the set is a `SINGLE` bucket and another entry arrives:
+ *      -> it is promoted to a `VECTOR` bucket (sorted by expiry).
+ * 3. If the `VECTOR` exceeds `VOLATILESET_VECTOR_BUCKET_MAX_SIZE` (127):
+ *      -> the set becomes a `RAX`, and existing entries are migrated.
+ * 4. IF the set is using RAX encoding it will locate a bucket to add the entry
+ *    following the strategy explained below.
+ *
+ *-----------------------------------------------------------------------------
+ * RAX Bucket and Dynamic Splitting
+ *-----------------------------------------------------------------------------
+ *
+ * Each bucket in the RAX bucket corresponds to a **time window**, defined by
+ * its bucket timestamp (`bucket_ts`). This timestamp represents the **END** of
+ * the time window. Entries in the bucket must expire *before* this timestamp.
+ *
+ * Time windows are defined in granular ranges:
+ *   - Minimum granularity: VOLATILESET_BUCKET_INTERVAL_MIN (16 ms)
+ *   - Maximum granularity: VOLATILESET_BUCKET_INTERVAL_MAX (8192 ms)
+ *
+ * A bucket can only contain entries that:
+ *   1. Have expiry < bucket_ts
+ *   2. Do not fit into any bucket with a smaller timestamp (i.e., earlier window)
+ *
+ * The structure allows multiple encodings:
+ *     VSET_BUCKET_SINGLE  - A single pointer to one entry.
+ *     VSET_BUCKET_VECTOR  - A sorted vector of pointers (up to 127 entries).
+ *     VSET_BUCKET_HT      - A hashtable used when vectors become too dense.
+ *
+ * Bucket Timestamp (END of window):
+ *
+ *        |------------------ Bucket Span ------------------|
+ *        [window_start .................................. bucket_ts)
+ *
+ * ASCII Layout Example:
+ *
+ *   Timeline:             ---> increasing time --->
+ *                         +------+---------+--------+
+ *                         | B0   |   B1    |   B2   |
+ *                         | ts=32| ts=128  | ts=2048|
+ *                         +------+---------+--------+
+ *                         ^      ^         ^
+ *                         |      |         |
+ *               [E1,E2] ∈ B0   [E3...E7] ∈ B1    [E8...] ∈ B2
+ *               All entries expire BEFORE their bucket_ts
```

Comment:
The three arrows are support to start on the three intervals? It's not quite visible now.

Making the buckets rectangles wider makes the intervals on the bottom line fit near the start of the arrows.

```suggestion
 *   Timeline:         ----------> increasing time ----------->
 *                     +--------------+-------------+---------+
 *                     | B0           | B1          |   B2    |
 *                     | ts=32        | ts=128      | ts=2048 |
 *                     +--------------+-------------+---------+
 *                     ^              ^             ^
 *                     |              |             |
 *           [E1,E2] ∈ B0      [E3...E7] ∈ B1     [E8...] ∈ B2
 *           All entries expire BEFORE their bucket_ts
```

Reply:
grrrr. thank you. will fix that